core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchAppRequestClient.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.stitch.core.internal.net;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.mongodb.stitch.core.StitchRequestErrorCode;
+import com.mongodb.stitch.core.StitchRequestException;
+import com.mongodb.stitch.core.internal.common.IoUtils;
+import com.mongodb.stitch.core.internal.common.StitchObjectMapper;
+
+import java.io.ByteArrayOutputStream;
+import java.nio.charset.StandardCharsets;
+
+public class StitchAppRequestClient extends StitchRequestClient {
+  private static final String BOOTSTRAP_ERROR_MESSAGE_INVALID_HOSTNAME =
+      "invalid hostname in metadata: %s";
+
+  private final String clientAppId;
+
+  private AppMetadata appMetadata;
+
+  public StitchAppRequestClient(
+      final String clientAppId,
+      final String baseUrl,
+      final Transport transport,
+      final Long defaultRequestTimeout
+  ) {
+    super(baseUrl, transport, defaultRequestTimeout);
+    this.clientAppId = clientAppId;
+  }
+
+  /**
+   * Performs a request against a Stitch app server determined by the deployment model
+   * of the underlying app. Throws a Stitch specific exception if the request fails.
+   *
+   * @param stitchReq the request to perform.
+   * @return a {@link Response} to the request.
+   */
+  @Override
+  public Response doRequest(final StitchRequest stitchReq) {
+    if (appMetadata == null) {
+      init(clientAppId);
+    }
+    // init throws if no location is available, so this should be safe
+
+    final Response response;
```

Comment:
[minor] Given this class extends `StitchRequestClient`, we should try to reuse its methods as much as possible to avoid the risk of missing out on logic that it has and we forget to implement here (and preserve bug compatibility), if it exists. To that end, I would add overloaded methods to `StitchRequestClient` for `doRequest` and `doStreamRequest` that take the URL discovered here. That way those methods can be very short stubs that end up calling the original `doRequest`/`doStreamRequest`.

Reply:
done